spring-cloud-function-context/src/main/java/org/springframework/cloud/function/cloudevent/CloudEventMessageUtils.java

```diff
@@ -272,6 +268,67 @@ else if (StringUtils.hasText(inputContentType)) {
 		return inputMessage;
 	}
 
+	/**
+	 * Attempts to {@link #canonicalizeHeaders canonicalize} the headers of a message.
+	 * @param message the message
+	 * @return a copy of the message with the canonicalized headers or the passed in unmodified message if no
+	 * headers were canonicalized
+	 */
+	// VisibleForTesting
+	static Message<?> canonicalizeHeadersWithPossibleCopy(Message<?> message) {
+		Map<String, Object> headers = new HashMap<>(message.getHeaders());
+		boolean headersModified = canonicalizeHeaders(headers, false);
+		if (headersModified) {
+			message = MessageBuilder.fromMessage(message)
+					.removeHeaders("*")
+					.copyHeaders(headers)
+					.build();
+		}
+		return message;
+	}
+
+	/**
```

Comment:
I regret moving this while modifying it at the same time. My bad. Here is what has changed.... 
- `modified` flag returned to caller